src/merkle/store/mod.rs

```diff
@@ -280,6 +277,37 @@ impl<T: KvMap<RpoDigest, StoreNode>> MerkleStore<T> {
         })
     }
 
+    /// Iterator over the non-empty leaves of the Merkle tree associated with the specified `root`
+    /// and `max_depth`.
+    pub fn non_empty_leaves(
+        &self,
+        root: RpoDigest,
+        max_depth: u8,
+    ) -> impl Iterator<Item = (NodeIndex, RpoDigest)> + '_ {
+        let empty_roots = EmptySubtreeRoots::empty_hashes(max_depth);
+        let mut stack = Vec::new();
+        stack.push((NodeIndex::new_unchecked(0, 0), root));
+
+        core::iter::from_fn(move || {
+            while let Some((index, node_hash)) = stack.pop() {
+                if index.depth() == max_depth {
+                    return Some((index, node_hash));
+                }
+
+                if let Some(node) = self.nodes.get(&node_hash) {
+                    if !empty_roots.contains(&node.left) {
+                        stack.push((index.left_child(), node.left));
+                    }
+                    if !empty_roots.contains(&node.right) {
+                        stack.push((index.right_child(), node.right));
+                    }
```

Comment:
One thing I haven't thought through enough: how should we think about empty leaves? I think we have 3 options:

1. Any root of an empty Merkle tree up to `max_depth` (this is how it is implemented now).
2. Root of empty an empty Merkle tree at depth equal to `index.depth`.
3. `EMPTY_WORD`.

I think the 3rd option would be more consistent with how we apply the diff. But also, I wonder if there are any real cases when option 1 and 2 would give different results.

Reply:
Ah - I think I misunderstood the code. I thought that if we see that a node belongs to the `empty_roots` set, we treat it as a leaf - but looking at it now, I think we are just skipping tree traversal in such cases.

Reply:
Yes the implication here is that if the node hash is in the `empty_roots` set then the leaves associated with this branch are empty and as such are not of interest of the iterator.